From b367711dc386acb156152db5bbdc8dd3410be085 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:09:27 -0400
Subject: [usb] usbtmc: fix DMA on stack

Message-id: <1410887463-20674-496-git-send-email-dzickus@redhat.com>
Patchwork-id: 95474
O-Subject: [RHEL7 PATCH 495/591] USB: usbtmc: fix DMA on stack
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit d846b7650db3fcca7901b6e23f6416c3601a3dfe
Author: Oliver Neukum <oneukum@suse.de>
Date:   Mon May 19 13:54:57 2014 +0200

    USB: usbtmc: fix DMA on stack

    send_request_dev_dep_msg_in() use a buffer allocated on the stack.
    Fix by kmalloc()ing the buffer.

    Signed-off-by: Oliver Neukum <oneukum@suse.de>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/usbtmc.c b/drivers/usb/class/usbtmc.c
index cfbec9c..103a6e9 100644
--- a/drivers/usb/class/usbtmc.c
+++ b/drivers/usb/class/usbtmc.c
@@ -383,9 +383,12 @@ exit:
 static int send_request_dev_dep_msg_in(struct usbtmc_device_data *data, size_t transfer_size)
 {
  int retval;
- u8 buffer[USBTMC_HEADER_SIZE];
+ u8 *buffer;
  int actual;
 
+ buffer = kmalloc(USBTMC_HEADER_SIZE, GFP_KERNEL);
+ if (!buffer)
+  return -ENOMEM;
  /* Setup IO buffer for REQUEST_DEV_DEP_MSG_IN message
   * Refer to class specs for details
   */
@@ -417,6 +420,7 @@ static int send_request_dev_dep_msg_in(struct usbtmc_device_data *data, size_t t
  if (!data->bTag)
   data->bTag++;
 
+ kfree(buffer);
  if (retval < 0) {
   dev_err(&data->intf->dev, "usb_bulk_msg in send_request_dev_dep_msg_in() returned %d\n", retval);
   return retval;
-- 
1.7.1